File: @archive/@deprecated/go_ai_service_2025_11_11/ai 2/messaging.go
package ai

import (
	"encoding/json"
	"fmt"
	"time"
)

/**
 * Unified Messaging System - 统一消息传递系统 (Go端)
 * 
 * Phase 46.14+: 三端统一消息显示
 * 
 * 与Rust/Python保持一致的消息格式
 */

// MessageType 消息类型
type MessageType string

const (
	MessageTypeInfo     MessageType = "info"
	MessageTypeWarning  MessageType = "warning"
	MessageTypeError    MessageType = "error"
	MessageTypeSuccess  MessageType = "success"
	MessageTypeProgress MessageType = "progress"
	MessageTypeStatus   MessageType = "status"
)

// MessageLevel 消息级别
type MessageLevel int

const (
	MessageLevelDebug    MessageLevel = 0
	MessageLevelInfo     MessageLevel = 1
	MessageLevelWarning  MessageLevel = 2
	MessageLevelError    MessageLevel = 3
	MessageLevelCritical MessageLevel = 4
)

// UnifiedMessage 统一消息结构
type UnifiedMessage struct {
	ID        string                 `json:"id"`
	Type      MessageType            `json:"type"`
	Level     MessageLevel           `json:"level"`
	Source    string                 `json:"source"`    // "go-ai"
	Component string                 `json:"component"`
	Message   string                 `json:"message"`
	Timestamp int64                  `json:"timestamp"`
	Progress  *int                   `json:"progress,omitempty"`   // 0-100
	Data      map[string]interface{} `json:"data,omitempty"`
	ErrorCode *string                `json:"error_code,omitempty"`
	TraceID   *string                `json:"trace_id,omitempty"`
}

// NewMessage 创建新消息
func NewMessage(msgType MessageType, level MessageLevel, component string, message string) *UnifiedMessage {
	return &UnifiedMessage{
		ID:        generateMessageID(),
		Type:      msgType,
		Level:     level,
		Source:    "go-ai",
		Component: component,
		Message:   message,
		Timestamp: time.Now().Unix(),
	}
}

// generateMessageID 生成消息ID
func generateMessageID() string {
	return fmt.Sprintf("msg_%x", time.Now().UnixNano())
}

// Info 创建信息消息
func NewInfoMessage(component string, message string) *UnifiedMessage {
	return NewMessage(MessageTypeInfo, MessageLevelInfo, component, message)
}

// Warning 创建警告消息
func NewWarningMessage(component string, message string) *UnifiedMessage {
	return NewMessage(MessageTypeWarning, MessageLevelWarning, component, message)
}

// Error 创建错误消息
func NewErrorMessage(component string, message string) *UnifiedMessage {
	return NewMessage(MessageTypeError, MessageLevelError, component, message)
}

// Success 创建成功消息
func NewSuccessMessage(component string, message string) *UnifiedMessage {
	return NewMessage(MessageTypeSuccess, MessageLevelInfo, component, message)
}

// Progress 创建进度消息
func NewProgressMessage(component string, message string, progress int) *UnifiedMessage {
	msg := NewMessage(MessageTypeProgress, MessageLevelInfo, component, message)
	if progress > 100 {
		progress = 100
	}
	msg.Progress = &progress
	return msg
}

// WithErrorCode 设置错误码
func (m *UnifiedMessage) WithErrorCode(code string) *UnifiedMessage {
	m.ErrorCode = &code
	return m
}

// WithTraceID 设置追踪ID
func (m *UnifiedMessage) WithTraceID(traceID string) *UnifiedMessage {
	m.TraceID = &traceID
	return m
}

// WithData 设置附加数据
func (m *UnifiedMessage) WithData(data map[string]interface{}) *UnifiedMessage {
	m.Data = data
	return m
}

// ToJSON 转换为JSON
func (m *UnifiedMessage) ToJSON() string {
	data, err := json.Marshal(m)
	if err != nil {
		return "{}"
	}
	return string(data)
}

// FromJSON 从JSON解析
func FromJSON(jsonStr string) (*UnifiedMessage, error) {
	var msg UnifiedMessage
	err := json.Unmarshal([]byte(jsonStr), &msg)
	return &msg, err
}

// Emit 发送消息（输出到stdout供其他端读取）
func (m *UnifiedMessage) Emit() {
	// 输出到stdout，前缀PIXLY_MSG:供解析
	fmt.Printf("PIXLY_MSG:%s\n", m.ToJSON())
	
	// 同时记录到日志系统
	switch m.Level {
	case MessageLevelDebug:
		Debug(m.Component, m.Message)
	case MessageLevelInfo:
		Info(m.Component, m.Message)
	case MessageLevelWarning:
		Warning(m.Component, m.Message)
	case MessageLevelError, MessageLevelCritical:
		Error(m.Component, m.Message)
	}
}

// MessageChannel已移除 - 过度设计，直接使用快捷函数

// 快捷函数

// SendInfo 发送信息消息
func SendInfo(component string, message string) {
	NewInfoMessage(component, message).Emit()
}

// SendWarning 发送警告消息
func SendWarning(component string, message string) {
	NewWarningMessage(component, message).Emit()
}

// SendError 发送错误消息
func SendError(component string, message string, errorCode *string) {
	msg := NewErrorMessage(component, message)
	if errorCode != nil {
		msg = msg.WithErrorCode(*errorCode)
	}
	msg.Emit()
}

// SendSuccess 发送成功消息
func SendSuccess(component string, message string) {
	NewSuccessMessage(component, message).Emit()
}

// SendProgress 发送进度消息
func SendProgress(component string, message string, progress int) {
	NewProgressMessage(component, message, progress).Emit()
}

// MessageReader已移除 - 当前未使用进程间通信

File: core/rust/src/messaging.rs
/**
 * Unified Messaging System - 统一消息传递系统
 * 
 * 🔥 Phase 46.14+: 三端统一消息显示
 * 
 * 功能：
 * - 统一消息格式（进度/状态/错误/通知）
 * - 跨端消息传递（Go ↔ Rust ↔ Python）
 * - 实时进度更新
 * - 多级别消息（Info/Warning/Error/Success）
 * 
 * @module messaging
 */
use serde::{Deserialize, Serialize};
use std::time::{SystemTime, UNIX_EPOCH};
// 🔧 统一日志系统
use tracing::{info, warn, error};

/// 消息类型
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum MessageType {
    /// 信息消息
    Info,
    /// 警告消息
    Warning,
    /// 错误消息
    Error,
    /// 成功消息
    Success,
    /// 进度更新
    Progress,
    /// 状态变更
    Status,
}

/// 消息级别
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
pub enum MessageLevel {
    /// 调试（开发用）
    Debug = 0,
    /// 普通信息
    Info = 1,
    /// 警告
    Warning = 2,
    /// 错误
    Error = 3,
    /// 严重错误
    Critical = 4,
}

/// 统一消息结构
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UnifiedMessage {
    /// 消息ID（唯一标识）
    pub id: String,
    
    /// 消息类型
    #[serde(rename = "type")]
    pub msg_type: MessageType,
    
    /// 消息级别
    pub level: MessageLevel,
    
    /// 来源（go-ai/rust-core/python-ml）
    pub source: String,
    
    /// 组件名称
    pub component: String,
    
    /// 消息内容
    pub message: String,
    
    /// 时间戳（Unix时间戳）
    pub timestamp: u64,
    
    /// 进度百分比（0-100，仅Progress类型）
    #[serde(skip_serializing_if = "Option::is_none")]
    pub progress: Option<u8>,
    
    /// 附加数据
    #[serde(skip_serializing_if = "Option::is_none")]
    pub data: Option<serde_json::Value>,
    
    /// 错误码（仅Error类型）
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error_code: Option<String>,
    
    /// 追踪ID（用于关联多条消息）
    #[serde(skip_serializing_if = "Option::is_none")]
    pub trace_id: Option<String>,
}

impl UnifiedMessage {
    /// 创建新消息
    pub fn new(
        msg_type: MessageType,
        level: MessageLevel,
        component: &str,
        message: String,
    ) -> Self {
        let id = Self::generate_id();
        let timestamp = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_secs();
        
        Self {
            id,
            msg_type,
            level,
            source: "rust-core".to_string(),
            component: component.to_string(),
            message,
            timestamp,
            progress: None,
            data: None,
            error_code: None,
            trace_id: None,
        }
    }
    
    /// 生成消息ID
    fn generate_id() -> String {
        use std::collections::hash_map::DefaultHasher;
        use std::hash::{Hash, Hasher};
        
        let now = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_micros();
        
        let mut hasher = DefaultHasher::new();
        now.hash(&mut hasher);
        
        format!("msg_{:x}", hasher.finish())
    }
    
    /// 创建信息消息
    pub fn info(component: &str, message: String) -> Self {
        Self::new(MessageType::Info, MessageLevel::Info, component, message)
    }
    
    /// 创建警告消息
    pub fn warning(component: &str, message: String) -> Self {
        Self::new(MessageType::Warning, MessageLevel::Warning, component, message)
    }
    
    /// 创建错误消息
    pub fn error(component: &str, message: String) -> Self {
        Self::new(MessageType::Error, MessageLevel::Error, component, message)
    }
    
    /// 创建成功消息
    pub fn success(component: &str, message: String) -> Self {
        Self::new(MessageType::Success, MessageLevel::Info, component, message)
    }
    
    /// 创建进度消息
    pub fn progress(component: &str, message: String, progress: u8) -> Self {
        let mut msg = Self::new(MessageType::Progress, MessageLevel::Info, component, message);
        msg.progress = Some(progress.min(100));
        msg
    }
    
    /// 设置错误码
    pub fn with_error_code(mut self, code: String) -> Self {
        self.error_code = Some(code);
        self
    }
    
    /// 设置追踪ID
    pub fn with_trace_id(mut self, trace_id: String) -> Self {
        self.trace_id = Some(trace_id);
        self
    }
    
    /// 设置附加数据
    pub fn with_data(mut self, data: serde_json::Value) -> Self {
        self.data = Some(data);
        self
    }
    
    /// 转换为JSON字符串
    pub fn to_json(&self) -> String {
        serde_json::to_string(self).unwrap_or_else(|_| "{}".to_string())
    }
    
    /// 从JSON字符串解析
    pub fn from_json(json: &str) -> Result<Self, serde_json::Error> {
        serde_json::from_str(json)
    }
    
    /// 发送到标准输出（供其他端读取）
    pub fn emit(&self) {
        // JSON格式输出到stdout，供Go/Python读取
        println!("PIXLY_MSG:{}", self.to_json());
        
        // 同时记录到日志
        match self.level {
            MessageLevel::Debug => {
                tracing::debug!("[{}] {}", self.component, self.message);
            }
            MessageLevel::Info => {
                info!("[{}] {}", self.component, self.message);
            }
            MessageLevel::Warning => {
                warn!("[{}] {}", self.component, self.message);
            }
            MessageLevel::Error | MessageLevel::Critical => {
                error!("[{}] {}", self.component, self.message);
            }
        }
    }
}

// MessageChannel已移除 - 过度设计，直接使用快捷函数即可

/// 快捷函数：发送信息消息
pub fn send_info(component: &str, message: String) {
    UnifiedMessage::info(component, message).emit();
}

/// 快捷函数：发送警告消息
pub fn send_warning(component: &str, message: String) {
    UnifiedMessage::warning(component, message).emit();
}

/// 快捷函数：发送错误消息
pub fn send_error(component: &str, message: String, error_code: Option<String>) {
    let mut msg = UnifiedMessage::error(component, message);
    if let Some(code) = error_code {
        msg = msg.with_error_code(code);
    }
    msg.emit();
}

/// 快捷函数：发送成功消息
pub fn send_success(component: &str, message: String) {
    UnifiedMessage::success(component, message).emit();
}

/// 快捷函数：发送进度消息
pub fn send_progress(component: &str, message: String, progress: u8) {
    UnifiedMessage::progress(component, message, progress).emit();
}
